Add support for Fractional limits and offsets (#88755)
ClickHouse/ClickHouse: docs/en/sql-reference/statements/select/limit.md

In the standard forms above, `n` and `m` are non-negative integers.

Negative limits are supported:

`LIMIT -m` selects the last `m` rows from the result.

`LIMIT -m OFFSET -n` selects the last `m` rows after skipping the last `n` rows. The `LIMIT -n, -m` syntax is equivalent.

> Selecting a fraction of the result is also supported:

Reviewer: ```suggestion
Moreover, selecting a fraction of the result is also supported:
```

> `LIMIT m` - If 0 < m < 1 then the first (m * 100)% percent from the result is returned.

Reviewer: ```suggestion
`LIMIT m` - if 0 < m < 1 then the first m * 100% of rows are returned.
```

> `LIMIT m` - if 0 < m < 1, then the first m * 100% of rows are returned.

Reviewer: You are right. My suggestion is not the actual behavior. Let's ignore it.

> `LIMIT m` - if 0 < m < 1, then the first m * 100% of rows are returned.

Reviewer: ```suggestion
`LIMIT m` - if 0 < m < 1, then the first `ceil(m × total_rows)` rows are returned.
```

What do you think of this?

> `LIMIT m` - if 0 < m < 1, then the first m * 100% of rows are returned.

Reviewer: Small grammar improvement:

`If a fractional number of rows resulted from the calculation, Its rounded up to the next whole number.`

to

`If a fractional number of rows results from the calculation, it is rounded up to the next whole number.`

`LIMIT m OFFSET n` - If 0 < m < 1 and 0 < n < 1 then (m * 100)% percent from the result is returned, after skipping the first (n * 100)%.

Examples:
    • `LIMIT 0.1` - selects the first 10% of the result.

>     • `LIMIT 0.5 OFFSET 1` - selects the median row.

Reviewer: ```suggestion
    • `LIMIT 1 OFFSET 0.5` - selects the median row.
```

    • `LIMIT 0.25 OFFSET 0.5` - selects 3rd quartile of the result.

> **Note**

> > • The fraction must be a Float64 number less than 1 and greater than zero.

Author: hyperlink to what?

> > • The fraction must be a Float64 number less than 1 and greater than zero.

Reviewer: Similar to:

```txt
- `value` — Input values, must be [Integer](../../data-types/int-uint.md) or [Float64](../../data-types/float.md) type.
```

`LIMIT -m OFFSET -n` selects the last `m` rows after skipping the last `n` rows. The `LIMIT -n, -m` syntax is equivalent.

Selecting a percentage from the table is also supported via **Fractions**:

`LIMIT 0.m` selects the first (m * 10)% percent of rows from the result.

`LIMIT 0.m OFFSET 0.n` selects the (m * 10)% percent of rows from the result, after skipping the first (n * 10)% percent of rows.

> 

Author: Tried to make it more clear, please re-review.

Examples:
    • `LIMIT 0.5 OFFSET 1` selects the median row
    • `LIMIT 0.5 OFFSET 0.25` selects 3rd quartile of the result

> **Note**

> > • Fractions must be Float32 numbers less than 1 and greater than zero

Reviewer: Why not `Float64`? Then user will be able to provide more precise the limit value.

Author: Done, but now with float64, `old_analyzer` has an imprecision of +1 or -1. I debugged it to ensure it's not a problem anywhere else, but I ensured it's just a rounding problem. This imprecision doesn't appear in `new_analyzer`, probably due to different handling of the constant node.

Example: `select * from numbers(10) limit 0.1` here, the `old_analyzer` will return 2 rows (an +1 imprecision), but as the row numbers grow, it will get identical to `new_analyzer`, which is the more precise one for some reason.

So in `03653_fractional_limit_offset.{sql,reference}` you can see that the `old_analyzer` results differ by +1,-1 from `new_analyzer`

Author: Done.
